typing/typecore.ml

```diff
@@ -5678,6 +5697,10 @@ let report_type_expected_explanation expl ppf =
       because "the left-hand side of a sequence"
   | When_guard ->
       because "a when-guard"
+  | Array_index ->
```

Comment:
Maybe add a test for both the `Array_index` and `String_index` case in the testsuite?